Reduce locking in HostClassLoader

hubertp opened the pull request:
### Pull Request Description

Using a neat trick described in https://github.com/enso-org/enso/pull/13219#discussion_r2131542150 to avoid large synchronization blocks.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Reduce locking in HostClassLoader

Using a neat trick described in https://github.com/enso-org/enso/pull/13219#discussion_r2131542150
to avoid large synchronization blocks.

Akirathan reviewed (commented):
This feels like something that could (should) be unit tested:
- Create an isolated `HostClassLoader` in the unit test
- Point it to some arbitrary jar archives / directories with classes
  - Scan from classpath system property, pass via sbt, ...
- Spawn multiple threads that will try to load same classes in parallel.

Or alternatively, move the scheduling/synchronization logic outside `HostClassLoader` and test it in isolation.

Let's try hard to ensure that we do not introduce hard-to-debug race conditions.

hubertp committed:
Merge branch 'develop' into wip/hubert/hostclassloader-improvements

hubertp commented:
Well, that wasn't a clean merge. Looks like tests got stuck :/

hubertp committed:
Properly propagate failed class loading in futures

If loading a classes would fail, then the result would not be properly
reported to the underlying future. As a result `future.get()` would get
stuck waiting infinitely.

This has happenned surprisingly a lot with nested classes, as a retry
mechanism is in place for replacing `.` with `$` in class name.
Also tests would hang infinitely.

hubertp committed:
Merge branch 'develop' into wip/hubert/hostclassloader-improvements

hubertp committed:
post-merge fmt

hubertp commented:
@JaroslavTulach checking if you are still OK with changes as I had to do a small modification and some time has passed.

hubertp commented:
> Let's try hard to ensure that we do not introduce hard-to-debug race conditions.

In theory, I agree. In practice this PR has been sitting delayed by that for a while as I didn't have an opportunity to do the refactoring. After a recent merge with develop our test suite has revealed some problems which are now fixed. I would argue we have sufficient enough coverage as-is for `HostClassLoader`.

JaroslavTulach commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/HostClassLoader.java:
```diff
@@ -63,49 +66,63 @@ public Class<?> loadClass(String name) throws ClassNotFoundException {
+      try {
+        placeholder[0].complete(loadClassUnsafe(name, resolve));
+      } catch (ClassNotFoundException e) {
+        placeholder[0].completeExceptionally(e);
```
- adding `completeExceptionally` in 4b8d3052e9e0cde6c20c062f76de1164fdc10589
- seems like a good idea

JaroslavTulach commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/HostClassLoader.java:
```diff
@@ -63,49 +66,63 @@ public Class<?> loadClass(String name) throws ClassNotFoundException {
+    try {
+      return pendingClass.get();
+    } catch (InterruptedException | ExecutionException e) {
+      throw new ClassNotFoundException("Unable to find class " + name, e);
```
- if the `Future` was `completeExceptionally` then
- the `e.getCause()` should be `instanceof ClassNotFoundException`, probably
- in such a case it might be better to throw the `e.getCause()` rather than creating a fresh exception, maybe?

JaroslavTulach reviewed (approved):
- I think the spirit of the change is still OK even after 4b8d3052e9e0cde6c20c062f76de1164fdc10589

> Let's try hard to ensure that we do not introduce hard-to-debug race conditions.

- all the hard synchronization is left on `ConcurrentHashMap` and `Future`
   - e.g. I am not afraid of race conditions, but deadlocks could appear...
- but there is one thing we should test:
   - have two classes that depend on each other (referencing a static field of the other for example)
   - block when one thread is about to load the first class and another thread is about to load the other class
   - let the threads continue and see if they finish
- if the test passes then this change is likely correct

hubertp committed:
Test multithreaded access to classloader

Added a test with mutual dependency between two classes.
The test also uses two threads to evaluate two code examples.
The setup should create sufficient race conditions.
If the test fails at any point, then we know that our logic is at fault.

hubertp commented:
> * but there is one thing we should test:

Fair enough. Added a test that should in theory create such a race condition.

JaroslavTulach committed:
Need both classes to start (and not finish) static class initialization at once

JaroslavTulach committed:
Initialization of the classes must depend on the other one

JaroslavTulach commented:
> Added a test that should in theory create such a race condition.

- the important thing to turn a theory into a practice is to block the two threads in the critical section
- that's done by 7e63f82
- once we have both classes "being classloaded at the same time"
- we need to create a dependency between them - done in 3a43d3f
- alas, now it deadlocks that means my original idea with a `Map<classname, Future>` was a bit too naive, sorry.
- on the other hand, not even regular `AppClassLoader` can handle such a wild classloading dependency as tested in b711de7
- so maybe we can remove my commits and integrate and hope normal classloading will be more sane than this test
- btw. the last case where I've seen this kind of problem for real was #7943 and #7850

JaroslavTulach committed:
Some timeouts to not block forever

JaroslavTulach committed:
Testing the same example with regular classloader

hubertp commented:
~~I'm not sure the two examples are comparable. 
The first fails because context never gets initialized - the usual `TruffleSafepoint` limitation we've encountered countless number of times:~~ Ignore that. This appears to only show up when executing an individual test from test class.
```
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@25.0.1/Native Method)
        - parking to wait for  <0x000000028bbbe228> (a com.oracle.truffle.api.impl.ThreadLocalHandshake$Barrier)
        at java.util.concurrent.locks.LockSupport.park(java.base@25.0.1/LockSupport.java:223)
...
```

Anyway, I'm leaning towards reverting your changes, as discussed offline.
The regular classloader test deadlock is, it seems, working according to the spec. [Here](https://stackoverflow.com/questions/53682182/class-initialization-deadlock-mechanism-explanation) is a good thread from SO.

hubertp committed:
Revert "Testing the same example with regular classloader"

This reverts commit b711de70138e3cda2bb0a85b2e19a851c7e2c5d7.

hubertp committed:
Revert "Some timeouts to not block forever"

This reverts commit 3de7b6becd3d9466e8cbd6352f59ec5fc32812b8.

hubertp committed:
Revert "Initialization of the classes must depend on the other one"

This reverts commit 3a43d3fad29d468f4e22a9d30a97e3187b8294fc.

hubertp committed:
Revert "Need both classes to start (and not finish) static class initialization at once"

This reverts commit 7e63f8277285783457bf101136694b21b09ab931.

JaroslavTulach commented:
> Anyway, I'm leaning towards reverting your changes, as discussed offline.

- OK, revert is reasonable
- Simpler way to revert is:
```
enso$ git checkout -f wip/hubert/hostclassloader-improvements
enso$ git revert 90b880a3ddfa6736602edd36d5d78235b58a6b51
enso$ git push -f
``` 
- but these individual commits will be gone when this PR is squashed, so it doesn't matter

hubertp committed:
Merge branch 'develop' into wip/hubert/hostclassloader-improvements

hubertp committed:
Merge branch 'develop' into wip/hubert/hostclassloader-improvements
